js/capture.js

```diff
@@ -3,51 +3,109 @@
 * third-party requests to storage.
 */
 const capture = {
+
   init() {
     this.addListeners();
   },
 
   addListeners() {
     // listen for each HTTP response
-    browser.webRequest.onResponseStarted.addListener(
-      (response) => this.sendThirdParty(response),
+    this.queue = [];
+    browser.webRequest.onResponseStarted.addListener((response) => {
+      const eventDetails = {
+        type: 'sendThirdParty',
+        data: response
+      };
+      this.queue.push(eventDetails);
+      this.processNextEvent();
+    },
       {urls: ['<all_urls>']});
     // listen for tab updates
     browser.tabs.onUpdated.addListener(
       (tabId, changeInfo, tab) => {
-        this.sendFirstParty(tabId, changeInfo, tab);
+        const eventDetails = {
+          type: 'sendFirstParty',
+          data: {
+            tabId: tabId,
+            changeInfo: changeInfo,
+            tab: tab
+          }
+        };
+        this.queue.push(eventDetails);
+        this.processNextEvent();
       });
   },
 
+  // Process each HTTP request or tab page load in order,
+  // so that async reads/writes to IndexedDB
+  // (via sendFirstParty and sendThirdParty) won't miss data
+  // The 'ignore' boolean ensures processNextEvent is only
+  // executed when the previous call to processNextEvent
+  // has completed.
+  async processNextEvent(ignore = false) {
+    if (this.processingQueue && !ignore) {
+      return;
+    }
+    if (this.queue.length >= 1) {
+      const nextEvent = this.queue.shift();
+      this.processingQueue = true;
+      switch (nextEvent.type) {
+        case 'sendFirstParty':
+          await this.sendFirstParty(
+            nextEvent.data.tabId,
+            nextEvent.data.changeInfo,
+            nextEvent.data.tab
+          );
+          break;
+        case 'sendThirdParty':
+          await this.sendThirdParty(nextEvent.data);
+          break;
+        default:
+          throw new Error(
+            'An event must be of type sendFirstParty or sendThirdParty.'
+          );
+      }
+      this.processNextEvent(true);
+    } else {
+      this.processingQueue = false;
+    }
+  },
+
   // Returns true if the request should be stored, otherwise false.
   shouldStore(tab) {
     const documentUrl = new URL(tab.url);
-    // ignore about:*, moz-extension:* & non-visible tabs (like dev tools)
+    // ignore about:* and moz-extension:*
     // also ignore third parties owned by first parties
```

Comment:
> Are we even checking for this any more?

I meant to remove the comment at line 78; removed!

> I would rather do more work in shouldStore than make it not encompass the whole shouldStore algo, this would make it really hard for a newcomer to understand.

`shouldStore` requires a `tab` object, which we can't get if the `tabId` is -1 (that was what was throwing an error [above](https://github.com/electrolyfish/lightbeam-we/pull/115#issuecomment-321410437)). How would we add non-visible tabs back in?